cpp/src/Slice/Parser.h

```diff
@@ -315,14 +316,17 @@ class Comment : public ::IceUtil::SimpleShared
 
     friend class Contained;
 };
-typedef ::IceUtil::Handle<Comment> CommentPtr;
+using CommentPtr = std::shared_ptr<Comment>;
 
 // ----------------------------------------------------------------------
 // GrammarBase
 // ----------------------------------------------------------------------
 
-class GrammarBase : public ::IceUtil::SimpleShared
+class GrammarBase : public virtual std::enable_shared_from_this<GrammarBase>
```

Comment:
It's not clear why the inheritance is virtual.

I would use move this inheritance to the derived classes that call shared_from_this (not GrammarBase), to avoid or reduce the dynamic_pointer_cast when calling shared_from_this()

Reply:
virtual is not required, but I don't think we can move inheritance to the derived classes, or we would end up with multiple inheritance from shared_from_this.